binding.cc

```diff
@@ -919,1155 +800,1353 @@ sn_crypto_stream_xor(
+  return crypto_hash_sha256_update(state_data, in.data(), in.size_bytes());
 }
 
-js_value_t *
-sn_crypto_hash_sha256_final(js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(2, crypto_hash_sha256_final)
-
-  SN_ARGV_BUFFER_CAST(crypto_hash_sha256_state *, state, 0)
-  SN_ARGV_TYPEDARRAY(out, 1)
+static inline int
+sn_crypto_hash_sha256_final(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state,
+  js_typedarray_span_t<> out
+) {
+  assert(state.size_bytes() == sizeof(crypto_hash_sha256_state));
+  auto state_data = reinterpret_cast<crypto_hash_sha256_state *>(state.data());
 
-  SN_THROWS(state_size != sizeof(crypto_hash_sha256_state), "state must be 'crypto_hash_sha256_STATEBYTES' bytes")
-  SN_ASSERT_LENGTH(out_size, crypto_hash_sha256_BYTES, "state")
+  assert(out.size_bytes() == crypto_hash_sha256_BYTES);
 
-  SN_RETURN(crypto_hash_sha256_final(state, out_data), "failed to finalise")
+  return crypto_hash_sha256_final(state_data, out.data());
 }
 
-js_value_t *
-sn_crypto_hash_sha512 (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(2, crypto_hash_sha512)
-
-  SN_ARGV_TYPEDARRAY(out, 0)
-  SN_ARGV_TYPEDARRAY(in, 1)
-
-  SN_ASSERT_LENGTH(out_size, crypto_hash_sha512_BYTES, "out")
+static inline int
+sn_crypto_hash_sha512(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> out,
+  js_typedarray_span_t<> in
+) {
+  assert(out.size_bytes() == crypto_hash_sha512_BYTES);
 
-  SN_RETURN(crypto_hash_sha512(out_data, in_data, in_size), "could not compute hash")
+  return crypto_hash_sha512(out.data(), in.data(), in.size_bytes());
 }
 
-js_value_t *
-sn_crypto_hash_sha512_init (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(1, crypto_hash_sha512_init)
-
-  SN_ARGV_BUFFER_CAST(crypto_hash_sha512_state *, state, 0)
-
-  SN_THROWS(state_size != sizeof(crypto_hash_sha512_state), "state must be 'crypto_hash_sha256_STATEBYTES' bytes")
+static inline int
+sn_crypto_hash_sha512_init(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state
+) {
+  assert(state.size_bytes() == sizeof(crypto_hash_sha512_state));
+  auto state_data = reinterpret_cast<crypto_hash_sha512_state *>(state.data());
 
-  SN_RETURN(crypto_hash_sha512_init(state), "failed to initialise sha512")
+  return crypto_hash_sha512_init(state_data);
 }
 
-js_value_t *
-sn_crypto_hash_sha512_update(js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(2, crypto_hash_sha512_update)
-
-  SN_ARGV_BUFFER_CAST(crypto_hash_sha512_state *, state, 0)
-  SN_ARGV_TYPEDARRAY(in, 1)
-
-  SN_THROWS(state_size != sizeof(crypto_hash_sha512_state), "state must be 'crypto_hash_sha256_STATEBYTES' bytes")
+static inline int
+sn_crypto_hash_sha512_update(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state,
+  js_typedarray_span_t<> in
+) {
+  assert(state.size_bytes() == sizeof(crypto_hash_sha512_state));
+  auto state_data = reinterpret_cast<crypto_hash_sha512_state *>(state.data());
```

Comment:
This could be improved by `libjstl` adding `js_typedarray_span_of_t<T>` similar to `js_arraybuffer_span_of_t<T>`.